Content.Server/GameObjects/Components/Doors/ServerDoorComponent.cs

```diff
@@ -310,139 +356,147 @@ public void TryClose(IEntity user)
-                var moles = adjacent.Air!.TotalMoles;
-                if (moles < minMoles)
-                    minMoles = moles;
-                if (moles > maxMoles)
-                    maxMoles = moles;
+                return true;
             }
 
-            return (maxMoles - minMoles) > threshold;
+            return accessReader.IsAllowed(user);
         }
 
-        public bool IsHoldingFire()
+        /// <summary>
+        /// Checks if we can close at all, for anyone or anything. Will return false if inhibited by an IDoorCheck component or if we are colliding with somebody while our Safety is on.
+        /// </summary>
+        /// <returns>Boolean describing whether this door can close.</returns>
+        public bool CanCloseGeneric()
         {
-            var atmosphereSystem = EntitySystem.Get<AtmosphereSystem>();
-
-            if (!Owner.Transform.Coordinates.TryGetTileAtmosphere(out var tileAtmos))
-                return false;
-
-            if (tileAtmos.Hotspot.Valid)
-                return true;
-
-            var gridAtmosphere = atmosphereSystem.GetGridAtmosphere(Owner.Transform.GridID);
-
-            foreach (var (_, adjacent) in gridAtmosphere.GetAdjacentTiles(tileAtmos.GridIndices))
+            if (Owner.TryGetComponent(out IDoorCheck? doorCheck) && !doorCheck.CloseCheck())
             {
-                if (adjacent.Hotspot.Valid)
-                    return true;
+                return false;
             }
 
-            return false;
+            return SafetyCheck();
         }
 
-        public bool Close()
+        /// <summary>
+        /// Checks if we are allowed to crush people, and if something is colliding with the door.
+        /// </summary>
+        /// <returns>True if we shouldn't worry about something colliding with us, false if we should.</returns>
+        public bool SafetyCheck()
         {
-            bool shouldCheckCrush = false;
-            if (Owner.TryGetComponent(out IPhysicsComponent? physics))
-                physics.CanCollide = true;
-
-            if (_canCrush && physics != null &&
-                physics.IsColliding(Vector2.Zero, false))
+            if (Safety && Owner.TryGetComponent(out IPhysicsComponent? physics))
             {
-                if (Safety)
+                // WE ARE NOT ACTUALLY BECOMING COLLIDEABLE. WE JUST NEED TO SEE IF WE SHOULD DELAY CLOSING DUE TO A PERSON.
+                var storedCanCollide = physics.CanCollide;
+                physics.CanCollide = true;
+                if (physics.IsColliding(Vector2.Zero, false))
                 {
-                    physics.CanCollide = false;
+                    physics.CanCollide = storedCanCollide;
                     return false;
                 }
-
-                // check if we crush someone while closing
-                shouldCheckCrush = true;
+                physics.CanCollide = storedCanCollide;
```

Comment:
You shouldn't set CanCollide multiple times like this because each one will trigger a broadphase update.